### PR TITLE
Refactor NavigationState.navigating tuple to a Struct

### PR DESCRIPTION
**Pull request by michaelkirk**
No behavioral change here, but I think this is a little nicer. Take it or leave it - no big deal.

### CODE REVIEW

File: apple/Sources/FerrostarCore/NavigationState.swift [hunk 1]
```diff
@@ -20,38 +20,4 @@ public struct NavigationState: Hashable {
         self.routeGeometry = routeGeometry
         self.isCalculatingNewRoute = isCalculatingNewRoute
     }
-
-    public var currentProgress: TripProgress? {
```
Comment (michaelkirk) [inline apple/Sources/FerrostarCore/NavigationState.swift:24]: At the callsite, we were using these methods like this:
```
              if case .navigating = navigationState?.tripState,
                        let progress = navigationState?.currentProgress
```

It was a bit of a shame to not be able to take advantage of the type information  — if it's a `.navigating` than we know `progress` isn't nil. Now with the structure we can capture that information and avoid the really verbose destructured tuple assignment.

**Comment by michaelkirk** [comment]
I'm sorry, but I don't immediately understand the CI failure.

https://github.com/stadiamaps/ferrostar/actions/runs/11187948122/job/31105894502?pr=286


```
Started: bash /Users/runner/work/_actions/stefanzweifel/git-auto-commit-action/v5/entrypoint.sh
INPUT_REPOSITORY value: .
INPUT_STATUS_OPTIONS: 
INPUT_FILE_PATTERN: apple/Sources/UniFFI/*
INPUT_BRANCH value: mkirk/refactor-tuple-struct
From https://github.com/stadiamaps/ferrostar
 * [new branch]      autocomplete-search -> origin/autocomplete-search
 * [new branch]      dependabot/gradle/android/agp-8.7.0 -> origin/dependabot/gradle/android/agp-8.7.0
 * [new branch]      feat/android/android-auto-init -> origin/feat/android/android-auto-init
 * [new branch]      feat/ios-speed-limit-view -> origin/feat/ios-speed-limit-view
 * [new branch]      main                -> origin/main
 * [new tag]         0.17.0              -> 0.17.0
fatal: invalid reference: mkirk/refactor-tuple-struct
Error: Invalid status code: 128
    at ChildProcess.<anonymous> (/Users/runner/work/_actions/stefanzweifel/git-auto-commit-action/v5/index.js:17:19)
    at ChildProcess.emit (node:events:519:28)
    at maybeClose (node:internal/child_process:1105:16)
    at ChildProcess._handle.onexit (node:internal/child_process:305:5) {
  code: 128
}
Error: Invalid status code: 128
    at ChildProcess.<anonymous> (/Users/runner/work/_actions/stefanzweifel/git-auto-commit-action/v5/index.js:17:19)
    at ChildProcess.emit (node:events:519:28)
    at maybeClose (node:internal/child_process:1105:16)
    at ChildProcess._handle.onexit (node:internal/child_process:305:5)
```

It appears that it's looking for my branch in your repository? That seems like it'd break for *any* foreign PR, but maybe I'm misreading.

**Comment by ianthetechie** [comment]
> I'm sorry, but I don't immediately understand the CI failure: https://github.com/stadiamaps/ferrostar/actions/runs/11187948122/job/31105894502?pr=286

@michaelkirk yeah, I know what happened. And I was wondering if we'd hit that error at some point 😂 Yay for untestable automation tools...

TL;DR make sure you run build_ios.sh, that Package.swift does not have local framework flags enabled, and that you've run swiftformat.

We have a CI action that automatically fixes these, but apparently this doesn't work cross-org even though you gave maintainers access to make changes.

File: apple/Sources/UniFFI/ferrostar.swift [hunk 2]
```diff
@@ -3465,57 +3465,94 @@ public enum TripState {
...
     /**
      * The navigation controller has reached the end of the trip.
      */
     case complete
 }
 
+/**
+ * State for when the navigation controller is actively navigating a trip.
+ */
+public struct TripNavigation: Equatable, Hashable {
```
[the remark this answers is not in the document]
Comment (ianthetechie) [reply apple/Sources/UniFFI/ferrostar.swift:3479]: I'm slightly confused here... Did you just edit this file by hand? Or did you forget to commit your Rust changes (and by extension, I assume Kotlin would have to change too)? This file is auto-generated by UniFFI.

[the remark this answers is not in the document]
Comment (michaelkirk) [reply apple/Sources/UniFFI/ferrostar.swift:3479]: Ohhh how embarrassing, I didn't realize this was auto-generated code. 

I might follow up with changes on the rust side, but I'll close this for now.

[the remark this answers is not in the document]
Comment (ianthetechie) [reply apple/Sources/UniFFI/ferrostar.swift:3479]: What's weird is that I could have swore this file used to have a header at the top warning that it was auto-generated. It seems to be gone now, and I'm not sure why 😂

References:

- https://github.com/mozilla/uniffi-rs/blob/8caac8e67c5782959d0c1b945ebfe489720c06c9/uniffi_bindgen/src/bindings/swift/templates/wrapper.swift#L4
- https://github.com/mozilla/uniffi-rs/issues/2256

[the remark this answers is not in the document]
Comment (michaelkirk) [reply apple/Sources/UniFFI/ferrostar.swift:3479]: TBH I probably wouldn't have noticed anyway given that I'm down on the 3,400th line.

That said, maybe the fact that I was editing a file with a billion lines should have been a clue that this was not a normal file. 🤔
